File: python/pia/resources/social.py
"""Official PIA SDK - Social Intelligence Resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from ..transport import AsyncTransport, SyncTransport
from ..types import SocialFeedResponse, SocialPostsResponse


def _social_params(
    *,
    symbol: Optional[str] = None,
    limit: Optional[int] = None,
    cursor: Optional[str] = None,
    before: Optional[str] = None,
    platform: Optional[str] = None,
    account: Optional[str] = None,
    q: Optional[str] = None,
) -> Dict[str, Any]:
    params: Dict[str, Any] = {}
    if symbol:
        params["symbol"] = symbol.strip().upper()
        params["q"] = symbol.strip().upper()
    if q:
        params["q"] = q
    if platform:
        params["platform"] = platform
    if account:
        params["account"] = account
    if before:
        params["before"] = before
    elif cursor:
        params["before"] = cursor
    if limit is not None:
        params["limit"] = limit
    return params


class SocialResource:
    """Synchronous Social Sentiment & Discussions resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_posts(
        self,
        *,
        symbol: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
        before: Optional[str] = None,
        platform: Optional[str] = None,
        account: Optional[str] = None,
        q: Optional[str] = None,
    ) -> SocialPostsResponse:
        """Fetches raw social posts using the backend's actual filters."""
        data = self._transport.request(
            "/api/v1/social/posts", method="GET",
            params=_social_params(symbol=symbol, limit=limit, cursor=cursor, before=before, platform=platform, account=account, q=q),
        )
        return SocialPostsResponse.from_dict(data)

    def get_feed(
        self,
        *,
        symbol: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> SocialFeedResponse:
        """Fetches curated social discussions and market sentiment posts."""
        data = self._transport.request(
            "/api/v1/social/feed", method="GET",
            params=_social_params(symbol=symbol, limit=limit, cursor=cursor),
        )
        return SocialFeedResponse.from_dict(data)


class AsyncSocialResource:
    """Asynchronous Social Sentiment & Discussions resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_posts(
        self,
        *,
        symbol: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
        before: Optional[str] = None,
        platform: Optional[str] = None,
        account: Optional[str] = None,
        q: Optional[str] = None,
    ) -> SocialPostsResponse:
        """Asynchronously fetches raw social posts using the backend's actual filters."""
        data = await self._transport.request(
            "/api/v1/social/posts", method="GET",
            params=_social_params(symbol=symbol, limit=limit, cursor=cursor, before=before, platform=platform, account=account, q=q),
        )
        return SocialPostsResponse.from_dict(data)

    async def get_feed(
        self,
        *,
        symbol: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> SocialFeedResponse:
        """Asynchronously fetches curated social discussions and market sentiment posts."""
        data = await self._transport.request(
            "/api/v1/social/feed", method="GET",
            params=_social_params(symbol=symbol, limit=limit, cursor=cursor),
        )
        return SocialFeedResponse.from_dict(data)

File: typescript/src/resources/social.ts
/**
 * Official PIA SDK - Social Intelligence API Resource
 */

import type { HttpTransport } from "../http/transport";
import type {
  GetSocialOptions,
  GetSocialPostsOptions,
  SocialFeedResponse,
  SocialPostsResponse,
} from "../types";

export class SocialResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Fetches the latest social sentiment posts from PostgreSQL / RSSHub ingestion pipeline.
   *
   * @param options Pagination limit, cursor, and optional symbol filter.
   */
  public async getPosts(options?: GetSocialPostsOptions): Promise<SocialPostsResponse> {
    const params = new URLSearchParams();
    if (options?.symbol) params.set("q", options.symbol.trim().toUpperCase());
    if (options?.limit) params.set("limit", String(options.limit));
    if (options?.cursor) params.set("before", options.cursor);
    if (options?.before) params.set("before", options.before);
    if (options?.platform) params.set("platform", options.platform);
    if (options?.account) params.set("account", options.account);
    if (options?.q) params.set("q", options.q);

    const query = params.toString() ? `?${params.toString()}` : "";
    return this.transport.request<SocialPostsResponse>(
      `/api/v1/social/posts${query}`,
      "GET",
      undefined,
      options
    );
  }

  /**
   * Fetches the real-time social discussion feed.
   */
  public async getFeed(options?: GetSocialOptions): Promise<SocialFeedResponse> {
    const params = new URLSearchParams();
    if (options?.symbol) params.set("symbol", options.symbol.trim().toUpperCase());
    if (options?.limit) params.set("limit", String(options.limit));
    if (options?.cursor) params.set("cursor", options.cursor);

    const query = params.toString() ? `?${params.toString()}` : "";
    return this.transport.request<SocialFeedResponse>(
      `/api/v1/social/feed${query}`,
      "GET",
      undefined,
      options
    );
  }
}
